pytorch_finufft/functional.py

```diff
@@ -836,3 +836,201 @@ def backward(  # type: ignore[override]
             None,
             None,
         )
+
+
+
+class finufft_type2(torch.autograd.Function):
+    """
+    FINUFFT 2D problem type 2
+    """
+
+    @staticmethod
+    def forward(
+        ctx: Any,
+        points: torch.Tensor,
+        targets: torch.Tensor,
+        out: Optional[torch.Tensor] = None,
+        finufftkwargs: Dict[str, Union[int, float]] = {},
+    ) -> torch.Tensor:
+        """
+        Evaluates the Type 2 NUFFT on the inputs.
+
+        NOTE: By default, the ordering is set to match that of Pytorch,
+         Numpy, and Scipy's FFT APIs. To match the mode ordering
+         native to FINUFFT, set fftshift=True.
+
+        ```
+        c[j] = SUM f[k1, k2] exp(+/-i (k1 x(j) + k2 y(j)))
+            k1, k2
+
+            for j = 0, ..., M-1, where the sum is over -N1/2 <= k1 <= (N1-1)/2,
+            -N2/2 <= k2 <= (N2-1)/2
+        ```
+
+        Parameters
+        ----------
+        ctx : Any
+            Pytorch context objecy
+        points_x : torch.Tensor
+            The non-uniform points x_j
+        points_y : torch.Tensor
+            The non-uniform points y_j
+        targets : torch.Tensor
+            The target Fourier mode coefficients f[k1, k2]
+        out : Optional[torch.Tensor], optional
+            Array to take the result in-place, by default None
+        fftshift : bool
+            If True centers the 0 mode in the resultant torch.Tensor, by default False
+        finufftkwargs : Dict[str, Union[int, float]]
+            Additional arguments will be passed into FINUFFT. See
+            https://finufft.readthedocs.io/en/latest/python.html. By default
+            an empty dictionary
+
+        Returns
+        -------
+        torch.Tensor
+            The resultant array c[j]
+
+        Raises
+        ------
+        ValueError
+            In the case of conflicting specification of the wave-mode ordering.
+        """
+
+        if out is not None:
+            print("In-place results are not yet implemented")
+
+        # TODO -- extend checks to 2d
+        checks._type2_checks(points, targets)
+
+        finufftkwargs = {k: v for k, v in finufftkwargs.items()}
+        _mode_ordering = finufftkwargs.pop("modeord", 1)
+        _i_sign = finufftkwargs.pop("isign", -1)
```

Comment:
Can you add a comment after each of these lines on whether or not it matches the FINUFFT default?